tiny_dnn/layers/layer.h

```diff
@@ -404,48 +404,94 @@ class layer : public node {
    * This can be used during training phase when parameter update is
    * required.
    *
-   * @param trainable_only flag to return only the trainable parameters.
+   * @param trainable_only flag to return only the trainable parameters
    * @return std::vector of pointers to parameters
    */
   Parameters parameters(bool trainable_only = false) {
     Parameters parameters;
     for (size_t i = 0; i < parameters_.size(); i++) {
       if (!trainable_only ||
-          (ith_parameter(i).is_trainable() && trainable_only)) {
-        parameters.push_back(&ith_parameter(i));
+          (parameter_at(i).is_trainable() && trainable_only)) {
+        parameters.push_back(&parameter_at(i));
       }
     }
     return parameters;
   }
 
-  /**
-   * @brief Get const pointers to parameters of this layer.
-   *
-   * This can be used during inference phase when parameter update is not
-   * required.
-   *
-   * @param trainable_only flag to return only the trainable parameters.
-   * @return const std::vector of const pointers to parameters
-   */
-  const ConstParameters parameters(bool trainable_only = false) const {
+  ConstParameters parameters(bool trainable_only = false) const {
     ConstParameters parameters;
     for (size_t i = 0; i < parameters_.size(); i++) {
       if (!trainable_only ||
-          (ith_parameter(i).is_trainable() && trainable_only)) {
-        parameters.push_back(&ith_parameter(i));
+          (parameter_at(i).is_trainable() && trainable_only)) {
+        parameters.push_back(&parameter_at(i));
       }
     }
     return parameters;
   }
 
-  Parameter &ith_parameter(size_t i) { return *parameters_[i]; }
+  /**
+   * @brief Get pointers to weights of this layer.
+   *
+   * @param trainable_only flag to return only the trainable weights
+   * @return std::vector of pointers to parameters which are weights
+   */
+  // todo (karandesai) : rename after old weights method is removed
+  Parameters weights_p(bool trainable_only = false) {
+    Parameters p = parameters(trainable_only);
+    for (size_t i = 0; i < p.size(); i++) {
+      if (p[i]->type() != parameter_type::weight) p.erase(p.begin() + i);
+    }
+    return p;
+  }
+
+  ConstParameters weights_p(bool trainable_only = false) const {
+    ConstParameters p = parameters(trainable_only);
+    for (size_t i = 0; i < p.size(); i++) {
+      if (p[i]->type() != parameter_type::weight) p.erase(p.begin() + i);
```

Comment:
this function is O(2n) while you could do it in O(n) if you specify in parameters the type you want. Besides, since there is no `weights(...)` I don't see the point of naming `weights_p(...)`.